go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -280,6 +280,183 @@ func (c *Controller) reconcile(name string) error {
+}
+
+// checkSubnetOverlaps validates that subnets from the current RA don't overlap
+// with any existing subnets in the same VRFs, both within the RA itself and
+// across different RAs
+func (c *Controller) checkSubnetOverlaps(ra *ratypes.RouteAdvertisements, selectedNetworks *selectedNetworks) error {
+	vrfSubnets := createVRFSubnetsMap(ra, selectedNetworks)
+
+	// Check for overlaps within each VRF
+	for vrf, subnets := range vrfSubnets {
+		// Check for overlaps within current RA
+		if err := checkWithinRAOverlaps(subnets); err != nil {
+			return fmt.Errorf("%w: error checking subnet overlap on RouteAdvertisement %q in VRF %q: %w", errConfig, ra.Name, vrf, err)
+		}
+		// Check for overlaps between current RA and other RAs in the same VRF
+		if err := c.checkCrossRAOverlaps(ra, vrf, subnets); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// checkWithinRAOverlaps checks if subnets within the current RA overlap with each other
+func checkWithinRAOverlaps(subnets []string) error {
+	for i, subnet1 := range subnets {
+		for j, subnet2 := range subnets {
+			if i >= j {
+				continue
+			}
+			overlap, err := subnetsOverlap(subnet1, subnet2)
+			if err != nil {
+				return fmt.Errorf("failed to check subnet overlap between %s and %s: %w", subnet1, subnet2, err)
+			}
+			if overlap {
+				return fmt.Errorf("overlapping CIDR detected: %s and %s", subnet1, subnet2)
+			}
+		}
+	}
+	return nil
+}
+
+// createVRFSubnetsMap creates a map of VRF -> subnets from selectedNetworks
+func createVRFSubnetsMap(ra *ratypes.RouteAdvertisements, selectedNetworks *selectedNetworks) map[string][]string {
+	vrfSubnets := make(map[string][]string)
+	if ra.Spec.TargetVRF == "auto" {
+		for _, vrf := range selectedNetworks.vrfs {
+			network := selectedNetworks.networkVRFs[vrf]
+			subnets := selectedNetworks.networkSubnets[network]
+			vrfSubnets[network] = subnets
+		}
+	} else {
+		vrfSubnets[ra.Spec.TargetVRF] = selectedNetworks.subnets
+	}
+	return vrfSubnets
+}
+
+// checkCrossRAOverlaps checks if subnets from current RA overlap with subnets from other RAs in the same VRF
+func (c *Controller) checkCrossRAOverlaps(ra *ratypes.RouteAdvertisements, targetVRF string, currentRASubnets []string) error {
+	allRAs, err := c.raLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, otherRA := range allRAs {
+		if otherRA.Name == ra.Name {
+			continue
+		}
+		otherSelectedNetworks, err := c.getSelectedNetworksForRA(otherRA)
+		if err != nil {
+			// If we can't get networks from another RA, log but don't fail
+			// This could happen if the other RA has configuration issues
+			klog.V(4).Infof("Failed to get selected networks from RouteAdvertisements %q: %v", otherRA.Name, err)
+			continue
+		}
+
+		otherVRFSubnets := createVRFSubnetsMap(otherRA, otherSelectedNetworks)
+		otherSubnets, hasMatchingVRF := otherVRFSubnets[targetVRF]
+		if !hasMatchingVRF || len(otherSubnets) == 0 {
+			continue
+		}
```

Comment:
instead of targetVRF we should pass to this function the "vrfSubnets" then we do some map's algebra by composing the key intersection between vrfSubnets and otherVRFSubnets

Then we interate that key set and extract subnets at vrfSubnets and otherVRFSubnets and to the overlap check.

So maybe we can use something like https://pkg.go.dev/v8.run/go/exp/util/maps to implement the key algebra

we can even do even more advanced stuff and instead of havimg map[string][]string have a map[string]Set[net.IPNet] and just implement the proper comparations (not sure about this though).

Reply:
I think it would be much simpler to do as you suggested and reuse subnetsSliceOverlap

Reply:
I don't think we need to go there. It's a simple search for string duplicates..